File: SinglyLinkedList.py
class Node:
    def __init__(self, element, next=None):
        self.element = element
        self.next = next

    def get_element(self):
        return self.element

    def get_next(self):
        return self.next

    def set_next(self, next_node):
        self.next = next_node

class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def size(self):
        return self.size

    def is_empty(self):
        return self.size == 0

    def first(self):
        if self.is_empty():
            return None
        return self.head.get_element()

    def last(self):
        if self.is_empty():
            return None
        return self.tail.get_element()

    def add_first(self, element):
        self.head = Node(element, self.head)
        if self.size == 0:
            self.tail = self.head
        self.size += 1

    def add_last(self, element):
        newest = Node(element, None)
        if self.is_empty():
            self.head = newest
        else:
            self.tail.set_next(newest)
        self.tail = newest
        self.size += 1

    def remove_first(self):
        if self.is_empty():
            return None
        answer = self.head.get_element()
        self.head = self.head.get_next()
        self.size -= 1
        if self.size == 0:
            self.tail = None
        return answer

    def __eq__(self, other):
        if other is None or not isinstance(other, SinglyLinkedList):
            return False
        if self.size != other.size:
            return False
        walk_a = self.head
        walk_b = other.head
        while walk_a is not None:
            if walk_a.get_element() != walk_b.get_element():
                return False
            walk_a = walk_a.get_next()
            walk_b = walk_b.get_next()
        return True

    def __copy__(self):
        other = SinglyLinkedList()
        if self.size > 0:
            other.head = Node(self.head.get_element(), None)
            walk = self.head.get_next()
            other_tail = other.head
            while walk is not None:
                newest = Node(walk.get_element(), None)
                other_tail.set_next(newest)
                other_tail = newest
                walk = walk.get_next()
        return other

    def __hash__(self):
        h = 0
        walk = self.head
        while walk is not None:
            h ^= hash(walk.get_element())
            h = (h << 5) | (h >> 27)
            walk = walk.get_next()
        return h

    def __str__(self):
        elements = []
        walk = self.head
        while walk is not None:
            elements.append(str(walk.get_element()))
            walk = walk.get_next()
        return '(' + ', '.join(elements) + ')'

    def remove_last(self):
        if self.is_empty():
            return None
        walk = self.head
        while walk.get_next() != self.tail:
            walk = walk.get_next()
        self.tail = walk
        walk.set_next(None)
        self.size -= 1

    def search(self, element):
        walk = self.head
        found = False
        while walk is not None:
            if walk.get_element() == element:
                found = True
            walk = walk.get_next()
        return found

    def concatenate(self, other_list):
        if self.is_empty():
            self.head = other_list.head
        else:
            self.tail.set_next(other_list.head)
        self.size += other_list.size
        self.tail = other_list.tail

if __name__ == "__main__":
    list1 = SinglyLinkedList()
    list1.add_first("MSP")
    list1.add_last("ATL")
    list1.add_last("BOS")
    list1.add_first("LAX")
    print(list1)
    list1.remove_last()
    print(list1)
    print(list1.search("LAM"))

    list2 = SinglyLinkedList()
    list2.add_first("YYZ")
    list2.add_first("YVR")
    print(list2)

    list1.concatenate(list2)
    print(list1)

File: SinglyLinkedList.java
public class SinglyLinkedList<E> implements Cloneable {

  private static class Node<E> {

    private E element;
    private Node<E> next;

    public Node(E e, Node<E> n) {
      element = e;
      next = n;
    }

    public E getElement() {
      return element;
    }

    public Node<E> getNext() {
      return next;
    }

    public void setNext(Node<E> n) {
      next = n;
    }
  }

  private Node<E> head = null;
  private Node<E> tail = null;
  private int size = 0;

  public SinglyLinkedList() {
  }

  public int size() {
    return size;
  }

  public boolean isEmpty() {
    return size == 0;
  }

  public E first() {
    if (isEmpty())
      return null;
    return head.getElement();
  }

  public E last() {
    if (isEmpty())
      return null;
    return tail.getElement();
  }

  public void addFirst(E e) {
    head = new Node<>(e, head);
    if (size == 0)
      tail = head;
    size++;
  }

  public void addLast(E e) {
    Node<E> newest = new Node<>(e, null);
    if (isEmpty())
      head = newest;
    else
      tail.setNext(newest);
    tail = newest;
    size++;
  }

  public E removeFirst() {
    if (isEmpty())
      return null;
    E answer = head.getElement();
    head = head.getNext();
    size--;
    if (size == 0)
      tail = null;
    return answer;
  }

  @SuppressWarnings({ "unchecked" })
  public boolean equals(Object o) {
    if (o == null)
      return false;
    if (getClass() != o.getClass())
      return false;
    SinglyLinkedList other = (SinglyLinkedList) o;
    if (size != other.size)
      return false;
    Node walkA = head;
    Node walkB = other.head;
    while (walkA != null) {
      if (!walkA.getElement().equals(walkB.getElement()))
        return false;
      walkA = walkA.getNext();
      walkB = walkB.getNext();
    }
    return true;
  }

  @SuppressWarnings({ "unchecked" })
  public SinglyLinkedList<E> clone() throws CloneNotSupportedException {
    SinglyLinkedList<E> other = (SinglyLinkedList<E>) super.clone();
    if (size > 0) {
      other.head = new Node<>(head.getElement(), null);
      Node<E> walk = head.getNext();
      Node<E> otherTail = other.head;
      while (walk != null) {
        Node<E> newest = new Node<>(walk.getElement(), null);
        otherTail.setNext(newest);
        otherTail = newest;
        walk = walk.getNext();
      }
    }
    return other;
  }

  public int hashCode() {
    int h = 0;
    for (Node walk = head; walk != null; walk = walk.getNext()) {
      h ^= walk.getElement().hashCode();
      h = (h << 5) | (h >>> 27);
    }
    return h;
  }

  public String toString() {
    StringBuilder sb = new StringBuilder("(");
    Node<E> walk = head;
    while (walk != null) {
      sb.append(walk.getElement());
      if (walk != tail)
        sb.append(", ");
      walk = walk.getNext();
    }
    sb.append(")");
    return sb.toString();
  }

  private void removeLast() {
    Node walk = head;
    while (walk.getNext() != tail) {
      walk = walk.getNext();
    }
    tail = walk;
    walk.setNext(null);
    size--;
  }

  private boolean search(String string) {
    Node walk = head;
    boolean found = false;
    while (walk != null) {
      if (walk.getElement().equals(string)) {
        found = true;
      }
      walk = walk.getNext();
    }
    return found;
  }

  private void concatenate(SinglyLinkedList<String> list2) {
    Node walk = head;
    while (walk.next != null) {
      walk = walk.getNext();
    }
    walk.setNext(list2.head);
    size = this.size + list2.size;
    tail = (Node<E>) list2.tail;
  }

  public static void main(String[] args) {
    SinglyLinkedList<String> list1 = new SinglyLinkedList<String>();
    list1.addFirst("MSP");
    list1.addLast("ATL");
    list1.addLast("BOS");
    
    list1.addFirst("LAX");
    System.out.println(list1);
    list1.removeLast();
    System.out.println(list1);
    System.out.println(list1.search("LAM"));

    SinglyLinkedList<String> list2 = new SinglyLinkedList<String>();
    list2.addFirst("YYZ");
    list2.addFirst("YVR");
    System.out.println(list2);

    list1.concatenate(list2);
    System.out.println(list1);
  }
}
